rfcs/0005-server-middleware-extensibility.md

```diff
@@ -0,0 +1,154 @@
+([RFC 0004](https://github.com/SAP/ui5-tooling/blob/master/rfcs/0004-simple-build-extensibility.md)), and translators.
+
+To separate "UI5 Projects" (i.e. things that represent UI5-artifacts for the browser) from tooling specific things
+like "extensions", an additional attribute "kind" is added to the ui5.yaml.
+
+A custom middleware will consist of at least a ui5.yaml defining it as an extension and
+a JavaScript implementation.
+
+A third party middleware which do not contain a ui5.yaml will be taken as it is. In this case the server expects
+the [default express API](https://expressjs.com/en/guide/writing-middleware.html). 
+
+#### Example middleware extension
+**`ui5.yaml`**:
+```yaml
+specVersion: "0.1"
+kind: extension
+type: middleware
+metadata:
+    name: myCustomMiddleware
+middleware:
+    path: middlewares/myCustomMiddleware.js
+```
+
+**`middlewares/myCustomMiddleware.js`**:
+```js
+module.exports = function({resourceCollections, tree}) {
+    return function (req, res, next) {
+      // middleware code...
+      next();
+    }
+};
+```
+
+#### Collecting and applying server middleware
+A middleware extension might be a standalone module or part of a project.
+
+If the extension is part of a project, the single `ui5.yaml` for the above example might look like this:
+
+```yaml
+specVersion: "0.1"
+kind: project
+type: application
+metadata:
+    name: my.application
+server:
+    customMiddleware:
+    - name: myCustomMiddleware
+      afterMiddleware: compression
+      configuration:
+        path: /myapp
+---
+specVersion: "0.1"
+kind: extension
+type: middleware
+metadata:
+    name: myCustomMiddleware
+middleware:
+    path: middlewares/myCustomMiddleware.js
+```
+
+In this case the extension is no dependency of any kind but automatically collected when the server is started.
+
+The `ui5-server` will detect the custom middleware configuration of the project my.application and inject the middleware
+into the servers start process to be available for the following HTTP requests.
+
+
+### Server middleware implementation
+A custom middleware implementation needs to return a function with the following signature (written in JSDoc):
+```js
+/**
+ * Custom ui5-server middleware example
+ *
+ * @param {Object} resourceCollections Contains the resource reader or collection to access project related files
+ * @param {AbstractReader} resourceCollections.source Reader to access the projects source files.
+ * @param {AbstractReader} resourceCollections.dependencies Reader or collecation to access the projects dependencies.
+ * @param {ReaderCollectionPrioritized} resourceCollections.combo Contains a prioritized collection to read and write all project files. 
+ * @returns {Function} Returns a middleware closure.
+ */
+module.exports = function({resourceCollections, tree}) {
```

Comment:
The configuration that can be supplied in the ui5.yaml is not passed to the middleware here.

In general I would prefer to keep the signature aligned with that of the custom tasks: https://github.com/SAP/ui5-project/blob/master/docs/BuildExtensibility.md#task-implementation

The project tree is currently missing there but is also not required for most server middlewares. We can discuss this in person later.

Reply:
What would be a reason for returning a promise for middleware functions?
Do you just want to align the signature with the tasks or do you see use cases where this might be required?
I would rather stick with our existing middleware signature as we would otherwise need to change our own middlewares in an incompatible way, or have them in a different signature which is confusing IMO.

You can still do async stuff in your middleware before accepting the first request.